### ISSUE TITLE
[Ch 04][Notebook][Major] FAQ Q6 방식 A 스니펫이 동작하지 않음 — single_label_classification 은 CE 매핑, num_labels=1 단독은 MSE

### ISSUE THREAD

**Issue by modesta-jihye-lee** [yoon-gu/neuqes-101#98]
## 대상

- **챕터**: Ch 04 — Binary on softmax
- **산출물**: Notebook (`04_softmax_binary/04_softmax_binary.ipynb` — `## ❓ FAQ` Q6 코드 스니펫, `cell018`)
- **기준 커밋**: `eed3dbe`
- **환경**: 정적 검토 — `transformers` 5.15.0 소스(`modeling_distilbert.py`)의 loss 매핑, Ch 10 실제 코드, CLAUDE.md `problem_type` 매핑표 대조. **Colab 전체 실행 완료** (2026-08-15 · CPU 런타임 · 45초 · 에러 0건)

## 실제 결과

FAQ Q6("Hugging Face `Trainer` 도 두 방식이 가능한가요?")의 **방식 A 스니펫이 동작하지 않습니다.**

```python
# 방식 A: num_labels=1, BCE 자동 적용            ← 주석이 사실과 다름
AutoModelForSequenceClassification.from_pretrained(
    "distilbert-base-uncased",
    num_labels=1,
    problem_type="single_label_classification",  # 또는 자동     ← 잘못된 인자
)
```

`transformers` 의 실제 매핑은 다음과 같습니다 (`models/distilbert/modeling_distilbert.py`, 5.15.0 기준 601-620행).

```python
if self.config.problem_type is None:
    if self.num_labels == 1:
        self.config.problem_type = "regression"          # ← "또는 자동" 은 MSELoss 가 됨
    ...
if   self.config.problem_type == "regression":                 loss_fct = MSELoss()
elif self.config.problem_type == "single_label_classification":
        loss_fct = CrossEntropyLoss()
        loss = loss_fct(logits.view(-1, self.num_labels), labels.view(-1))   # ← BCE 아님
elif self.config.problem_type == "multi_label_classification": loss_fct = BCEWithLogitsLoss()
```

세 가지가 어긋납니다.

| 스니펫의 주장 | 실제 |
|---|---|
| `problem_type="single_label_classification"` → **BCE 자동 적용** | **`CrossEntropyLoss`** 가 적용됨 |
| `# 또는 자동` (인자를 빼도 된다) | `num_labels=1` 단독은 **회귀로 추정 → `MSELoss`** |
| 이대로 쓰면 방식 A(sigmoid+BCE) 가 된다 | `num_labels=1` + CE 는 **클래스가 하나뿐인 softmax** — `log_softmax` 가 항상 0 이라 loss·gradient 가 0 으로 죽어 **학습 자체가 되지 않음** |

에러가 나지 않고 조용히 학습만 안 되는 형태라 학습자가 원인을 찾기 어렵습니다.

### 이 커리큘럼의 정답은 이미 Ch 10 에 있습니다

`10_bert_binary_sigmoid/10_bert_binary_sigmoid.ipynb`:

```python
model = AutoModelForSequenceClassification.from_pretrained(
    "distilbert-base-uncased",
    num_labels=1,
    problem_type="multi_label_classification",   # ← BCEWithLogitsLoss 자동 매핑
)
```

CLAUDE.md 의 매핑표도 같습니다 — `multi_label_classification` → `BCEWithLogitsLoss`, 라벨은 float 텐서.

## 기대 결과

```python
# 방식 A: num_labels=1 + BCEWithLogitsLoss (Ch 10 과 동일한 설정)
AutoModelForSequenceClassification.from_pretrained(
    "distilbert-base-uncased",
    num_labels=1,
    problem_type="multi_label_classification",  # num_labels=1 이어도 BCE 매핑에 이 값을 씁니다
)
# 라벨은 float 텐서여야 합니다 (0.0 / 1.0)

# 방식 B: num_labels=2, CE 자동 적용 (BERT 표준)
AutoModelForSequenceClassification.from_pretrained(
    "distilbert-base-uncased",
    num_labels=2,
)
```

`problem_type` 이름이 직관과 어긋나는 부분(`num_labels=1` 인데 `multi_label_...`)은 한 줄 각주로 짚어 주면 학습자가 덜 헷갈립니다. 방식 B 스니펫은 현재 그대로 정확합니다.

## 재현 절차

```bash
python3 - <<'PY'
import transformers, os, re
p = os.path.join(os.path.dirname(transformers.__file__),
                 "models/distilbert/modeling_distilbert.py")
src = open(p).read()
i = src.find("problem_type is None")
print(src[i-40:i+900])
PY

grep -n -A5 "AutoModelForSequenceClassification.from_pretrained" \
     10_bert_binary_sigmoid/10_bert_binary_sigmoid.ipynb | head
```

## 로그

```
transformers 5.15.0

601:  if self.config.problem_type is None:
602:      if self.num_labels == 1:
603:          self.config.problem_type = "regression"
615:  elif self.config.problem_type == "single_label_classification":
616:      loss_fct = CrossEntropyLoss()
618:  elif self.config.problem_type == "multi_label_classification":
619:      loss_fct = BCEWithLogitsLoss()

Ch 10 실제 코드: num_labels=1, problem_type="multi_label_classification"
```

## 전파 범위

| 산출물 | 상태 |
|---|---|
| `04_softmax_binary/04_softmax_binary.ipynb` `cell018` | ❌ |
| `pages/04-softmax_binary-wrapup.md` (FAQ Q6) | ❌ |
| `book/chapters/ch04_softmax_binary.tex` | ❌ → EPUB 영향 |

## 심각도

**Major** — 계획서 §9 의 "잘못된 코드" 에 해당합니다. 학습자가 복사해 쓰면 에러 없이 학습만 실패하고, 같은 커리큘럼의 Ch 10 코드와도 어긋납니다. 본문 흐름이 아닌 FAQ 스니펫이라 Blocker 는 아니며, 앞선 사례 #65(FAQ 코드 스니펫 오류 → Major)와 같은 등급으로 잡았습니다.

## 재실행 여부

마크다운 셀 수정이므로 **노트북 재실행은 불필요**합니다. WikiDocs 재변환 → EPUB 재빌드가 필요합니다.

## 담당

- 수정 담당자: 미정
- 재검수자: 등록자 (@modesta-jihye-lee)
- 연결: 챕터 검수 이슈 #28

**Comment by yoon-gu** [comment]
PASS — FAQ Q6 방식 A 스니펫이 에러 없이 학습만 실패하던 문제가 해소됐습니다. 노트북 `cell018` 이 제안하신 기대 결과와 일치하게 교체됐고 pages·tex 에도 전파됐습니다.

— PR #289 (master `6986584`, 2026-09-18). `origin/master` 기준으로 노트북·`executed/`·`pages/`·`tex` 전파까지 확인했습니다.
